diff -urN linux-2.6.23.9-orig/scripts/mkcompile_h linux-2.6.23.9/scripts/mkcompile_h
--- linux-2.6.23.9-orig/scripts/mkcompile_h	2010-01-20 01:32:00.000000000 +0800
+++ linux-2.6.23.9/scripts/mkcompile_h	2010-01-20 17:55:43.000000000 +0800
@@ -42,6 +42,9 @@
 export LANG=en_US
 SVN_VER=$(svn info ../../.. | grep "Rev:" | tr -d [:blank:][:alpha:]:)
 UTS_VERSION="r$SVN_VER"
+FW_VER=$(grep "CONFIG_WEBROCX_RELEASE" ../../../.config | tr -d [\"]: | awk -F"=" '{printf $2}')
+echo -e "Revision: ${SVN_VER:-unknown}\nBuild_Date: $TIMESTAMP\nFW_Version: $FW_VER" > ../../../bin/RVERSION
+
 CONFIG_FLAGS=""
 if [ -n "$SMP" ] ; then CONFIG_FLAGS="SMP"; fi
 if [ -n "$PREEMPT" ] ; then CONFIG_FLAGS="$CONFIG_FLAGS PREEMPT"; fi
